File: booksy-api/src/main/java/com/booksy/backend/controller/CartController.java
package com.booksy.backend.controller;

import com.booksy.backend.model.CartItem;
import com.booksy.backend.repository.CartItemRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/cart")
@CrossOrigin(origins = "*")
public class CartController {

    @Autowired
    private CartItemRepository cartItemRepository;

    // comision del 10%
    private static final double COMMISSION_RATE = 0.10;

    // obtener items del carrito de un usuario
    @GetMapping("/user/{userId}")
    public ResponseEntity<List<CartItem>> getCartItems(@PathVariable String userId) {
        List<CartItem> items = cartItemRepository.findByUserId(userId);
        return ResponseEntity.ok(items);
    }

    // agregar item al carrito
    @PostMapping
    public ResponseEntity<CartItem> addToCart(@RequestBody CartItem cartItem) {
        CartItem savedItem = cartItemRepository.save(cartItem);
        return ResponseEntity.status(HttpStatus.CREATED).body(savedItem);
    }

    // actualizar cantidad de un item
    @PutMapping("/{id}")
    public ResponseEntity<?> updateCartItem(@PathVariable String id, @RequestBody Map<String, Integer> request) {
        var itemOpt = cartItemRepository.findById(id);

        if (itemOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body("item no encontrado");
        }

        CartItem item = itemOpt.get();
        item.setQuantity(request.get("quantity"));

        CartItem updatedItem = cartItemRepository.save(item);
        return ResponseEntity.ok(updatedItem);
    }

    // eliminar item del carrito
    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteCartItem(@PathVariable String id) {
        if (!cartItemRepository.existsById(id)) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body("item no encontrado");
        }

        cartItemRepository.deleteById(id);
        return ResponseEntity.ok("item eliminado");
    }

    // calcular total del carrito con comision
    @GetMapping("/total/{userId}")
    public ResponseEntity<Map<String, Double>> getCartTotal(@PathVariable String userId) {
        List<CartItem> items = cartItemRepository.findByUserId(userId);

        // calcular subtotal
        double subtotal = items.stream()
                .mapToDouble(item -> item.getPricePerUnit() * item.getQuantity())
                .sum();

        // calcular comision
        double commission = subtotal * COMMISSION_RATE;

        // total final
        double total = subtotal + commission;

        Map<String, Double> response = new HashMap<>();
        response.put("subtotal", subtotal);
        response.put("commission", commission);
        response.put("total", total);

        return ResponseEntity.ok(response);
    }

    // limpiar carrito
    @DeleteMapping("/clear/{userId}")
    public ResponseEntity<?> clearCart(@PathVariable String userId) {
        cartItemRepository.deleteByUserId(userId);
        return ResponseEntity.ok("carrito limpiado");
    }
}

File: booksy-api/src/cart/cart.controller.ts
import { Controller, Get, Post, Put, Delete, Body, Param, HttpStatus, HttpException } from '@nestjs/common';
import { CartService } from './cart.service';

@Controller('api/cart')
export class CartController {
  constructor(private readonly cartService: CartService) {}

  // obtener items del carrito
  @Get('user/:userId')
  async getCartItems(@Param('userId') userId: string) {
    return this.cartService.findByUserId(userId);
  }

  // agregar item al carrito
  @Post()
  async addToCart(@Body() cartItemData: any) {
    return this.cartService.create(cartItemData);
  }

  // actualizar cantidad
  @Put(':id')
  async updateCartItem(@Param('id') id: string, @Body() body: { quantity: number }) {
    const item = await this.cartService.updateQuantity(id, body.quantity);
    if (!item) {
      throw new HttpException('item no encontrado', HttpStatus.NOT_FOUND);
    }
    return item;
  }

  // eliminar item
  @Delete(':id')
  async deleteCartItem(@Param('id') id: string) {
    const result = await this.cartService.delete(id);
    if (!result) {
      throw new HttpException('item no encontrado', HttpStatus.NOT_FOUND);
    }
    return { message: 'item eliminado' };
  }

  // calcular total con comision
  @Get('total/:userId')
  async getCartTotal(@Param('userId') userId: string) {
    return this.cartService.calculateTotal(userId);
  }

  // limpiar carrito
  @Delete('clear/:userId')
  async clearCart(@Param('userId') userId: string) {
    await this.cartService.clearCart(userId);
    return { message: 'carrito limpiado' };
  }
}
